### PR TITLE
Limit the permission of secret used by proxy cache service

### LINKED ISSUE #12257: The secret for the proxy to push layer should have minimal permission.
The PR #12214 makes the proxy secret can do anything, we should make sure the secret is short lived and it has minimal permission, i.e. push to a project that has proxy capability, and it expires quickly.

### PR DESCRIPTION
**Pull request by ywk253100**
Limit the permission of secret used by proxy cache service, fixes #12257

Signed-off-by: Wenkai Yin <yinw@vmware.com>

### CODE REVIEW

File: src/pkg/proxy/secret/authorizer.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	return &authorizer{}
+}
+
+type authorizer struct{}
+
+func (s *authorizer) Modify(req *http.Request) error {
+	if req == nil {
+		return errors.New("the request is null")
+	}
+	repository, _, ok := lib.MatchManifestURLPattern(req.URL.Path)
```
Comment (stonezdj) [inline src/pkg/proxy/secret/authorizer.go:41]: The nested if condition can be refactored to 
```
	if _, _, ok := MatchFunc1(); ok {

	}else if _, _, ok := MatchFunc2(); ok {

	}else if _, _, ok := MatchFunc3(); ok {

	}else {
		//not handler
		return
	}
	//handle it
```

File: src/lib/patterns.go [hunk 2]
```diff
@@ -29,3 +29,33 @@ var (
...
+	strs := V2BlobURLRe.FindStringSubmatch(path)
+	if len(strs) < 3 {
+		return "", "", false
+	}
+	return strs[1], strs[2], true
+}
+
+// MatchBlobUploadURLPattern checks whether the provided path matches the blob upload URL pattern,
+// if does, returns the repository as well
+func MatchBlobUploadURLPattern(path string) (repository string, match bool) {
```
Comment (stonezdj) [inline src/lib/patterns.go:55]: No need to return the repo, because the logic is already implement in artifact_info.middleware

Comment (ywk253100) [reply src/lib/patterns.go:55]: This is common utility function now.
And there is no artifact info when trying to push the artifact to local in the client side

File: src/pkg/proxy/secret/authorizer.go (hunk 1 above)
Comment (ywk253100) [reply src/pkg/proxy/secret/authorizer.go:41]: This authorizer is used when trying to push the artifact, this is in the client side, no artifact info in the context

**Pull request merged**